lib/matplotlib/figure.py

```diff
@@ -1523,6 +1524,190 @@ def subplots(self, nrows=1, ncols=1, sharex=False, sharey=False,
+            -------
+            2D object array
+
+            """
+            r0, *rest = inp
+            for j, r in enumerate(rest, start=1):
+                if len(r0) != len(r):
+                    raise ValueError(
+                        "All of the rows must be the same length, however "
+                        f"the first row ({r0!r}) has length {len(r0)} "
+                        f"and row {j} ({r!r}) has length {len(r)}."
+                    )
+            out = np.zeros((len(inp), len(r0)), dtype=object)
+            for j, r in enumerate(inp):
+                for k, v in enumerate(r):
+                    out[j, k] = v
+            return out
+
+        def _identify_keys_and_nested(layout):
+            """
+            Given a 2D object array, identify unique IDs and nested layouts
+
+            Parameters
+            ----------
+            layout : 2D numpy object array
+
+            Returns
+            -------
+            unique_ids : Set[object]
+                The unique non-sub layout entries in this layout
+            nested : Dict[Tuple[int, int]], 2D object array
+            """
+            unique_ids = set()
+            nested = {}
+            for j, row in enumerate(layout):
+                for k, v in enumerate(row):
+                    if v == empty_sentinel:
+                        continue
+                    elif not cbook.is_scalar_or_string(v):
+                        nested[(j, k)] = _make_array(v)
+                    else:
+                        unique_ids.add(v)
+
+            return unique_ids, nested
+
+        def _do_layout(gs, layout, unique_ids, nested):
+
+            rows, cols = layout.shape
+            output = dict()
+
+            for name in unique_ids:
+                if name == empty_sentinel:
+                    continue
+
+                indx = np.argwhere(layout == name)
+                start_row, start_col = np.min(indx, axis=0)
+                end_row, end_col = np.max(indx, axis=0) + 1
+                slc = (slice(start_row, end_row), slice(start_col, end_col))
+
+                if (layout[slc] != name).any():
+                    raise ValueError(
+                        f"While trying to layout\n{layout!r}\n"
+                        f"we found that the label {name!r} specifies an "
+                        "non-rectangular or non-contiguous area.")
+
+                if hasattr(name, 'set_subplotspec'):
+                    ax = name
+                    ax.set_subplotspec(gs[slc])
+                    ax.update_params()
+                    ax.set_position(ax.figbox)
+                    output[id(ax)] = ax
+                else:
+                    output[name] = self.add_subplot(
+                        gs[slc], **{'label': str(name), **subplot_kw})
+
+            for (j, k), nested_layout in nested.items():
+                rows, cols = nested_layout.shape
+                gs_n = gs[j, k].subgridspec(rows, cols, **gridspec_kw)
```

Comment:
are j and k ranges here?

Reply:
No, these are strictly integers (for now, see lines 1650-1651 just above this).  However, I do see how it could be extended to support slices.  I'm inclined to wait for someone to request that before we implement it as I think it might start to be harder to use this short hand than to just use GridSpec if you need to define subgridspecs over slices.